[로또 - 웹, DB] 슬로스 미션 제출합니다.

soojinroh opened the pull request.

indiveloper committed:
feat 기능 구현 목록 작성

indiveloper committed:
feat LottoNumber 클래스 구현

1부터 45까지의 로또 번호 하나 하나를 의미하는 LottoNumber 클래스 작성. 1부터 45까지의 숫자만 허용하고 이외의 숫자가 입력될 경우 예외를 발생시킴.

indiveloper committed:
refactor EXCEPTION_MESSAGE 문구 수정

EXCEPTION_MESSAGE 문구 수정

indiveloper committed:
feat LottoNumbers 클래스 구현

1부터 45까지의 값을 가지고 있는 LottoNumber 객체 45개를 할당 받은 Map의 일급 컬렉션. static으로 구현하여 프로그램 내에서 재사용 할 수 있게 함.

indiveloper committed:
feat AutoLottoGenerator 구현

랜덤한 LottoNumber객체 6개로 이루어진 List<LottoNumber>를 생성해 반환하는 AutoLottoGenerator 구현

indiveloper committed:
feat ManualLottoGenerator 구현

AutoLottoGenerator는 랜덤하게 번호를 선택해 List<LottoNumber>를 생성한 반면, ManualLottoGenerator는 사용자로부터 입력 받은 6개의 중복되지 않는 int로 이루어진 List를 입력 받아 List<LottoNumber>를 생성해 리턴함

indiveloper committed:
refactor 변수명 변경

allLottoNumbers를 lottoNumbers로 변수명 변경

indiveloper committed:
feat Money 클래스 구현

로또 구입 금액을 의미하는 Money 클래스 구현.
1,000원 이상 100,000원 이하의 금액만 입력 가능하다.
1,000원을 기준으로 입력해야 한다. 100원, 10원, 1원 단위가 입력되었을 경우 예외를 던진다.

indiveloper committed:
refactor 변수명 변경

indiveloper committed:
refactor 클래스명 변경 및 equals, hashCode 오버라이드

indiveloper committed:
feat Lotto 클래스 구현

- 로또 한 게임(6개의 번호를 가지고 있는, 1000원짜리)을 의미하는 Lotto 클래스 구현
- isContained()메서드를 이용해 중복하는 번호가 있는지 확인

indiveloper committed:
refactor 변수명 변경

indiveloper committed:
refactor update README.md

기능 구현 목록 수정

indiveloper committed:
feat Validator 클래스 구현

InputView에서 입력 값의 유효성을 확인하는 메서드들을 Validator 클래스로 이동시킴.
Money 클래스를 테스트하는 중에 유효하지 않은 값을 테스트하려고 하는데 이미 유효성 확인을 마친 값을 입력받는 다는 전제하에 Money 클래스를 구현하다보니, Money 클래스 내에서는 유효성을 검사하는 기능이 없었음. 그래서 Validator 클래스에 checkInputMoney() 메서드를 만들어 InputView에서도 유효성 검사를 하고 Money에서도 유효성 검사를 할 수 있게 수정함.

indiveloper committed:
refactor LottoGenerator return 타입 수정

List<Number>를 반환하던 것에서 Lotto를 만들어 반환하는 것으로 수정

                                                                                                                                                                                                                               InputView에서 입력 값의 유효성을 확인하는 메서드들을 Validator 클래스로 이동시킴.
                                                                                                                                                                                                                               Money 클래스를 테스트하는 중에 유효하지 않은 값을 테스트하려고 하는데 이미 유효성 확인을 마친 값을 입력받는 다는 전제하에 Money 클래스를 구현하다보니, Money 클래스 내에서는 유효성을 검사하는 기능이 없었음. 그래서 Validator 클래스에 checkInputMoney() 메서드를 만들어 InputView에서도 유효성 검사를 하고 Money에서도 유효성 검사를 할 수 있게 수정함.

indiveloper committed:
refactor 변수명 변경

생성자의 매개변수 List<Number>의 변수명을 lotto에서 lottoNumbers로  변경

indiveloper committed:
refactor 유효성 검사를 Validator클래스를 이용해 하는 것으로 수정

Money 클래스에서 입력 값에 대해 유효성 검사를 하던 것을 Validator 클래스로 옮겨 Validator 클래스에서 함수를 호출해 유효성 검사 하는 것으로 수정

indiveloper committed:
feat PurchasedLottos 클래스 구현

사용자가 구입한 모든 로또의 리스트 List<Lotto>를 가지고 있는 클래스

indiveloper committed:
feat WinningLotto 클래스 구현

지난주 당첨 번호와 보너스 볼 넘버를 가지고 있는 WinningLotto 클래스 구현

indiveloper committed:
refactor 패키지 수정

model 패키지 생성

indiveloper committed:
feat Main 클래스 구현

로또 게임을 진행할 Main 클래스 구현

indiveloper committed:
feat 로또 구입 결과 출력

콘솔 출력을 담당하는 OutputView 클래스에 로또 구입 내역을 출력하는 printPurchasedLottoResult() 함수 구현

indiveloper committed:
refactor toString 오버라이드

Lotto의 정보를 알 수 있게끔 toString 오버라이드

indiveloper committed:
refactor 하드코딩한 값 상수화

하드코딩한 값을 상수화 함

indiveloper committed:
refactor printInputLottoNumberMessage 분리

수동 구매 로또 입력 메시지와 당첨 로또 입력 메시지를 한 번에 처리하던 printInputLottoNumberMessage()를 분리하여 수동 로또를 구매할 때와 지난주 당첨 번호를 입력할 때 각각 다른 메서드를 호출해서 실행하게 수정

indiveloper committed:
feat 보너스 번호 중복 체크 기능 추가

보너스 번호가 이미 입력 받은 지난주 당첨번호와 중복되는지 확인하는 기능 추가

indiveloper committed:
feat 구입한 모든 로또를 의미하는 PurchasedLottos 클래스 구현

수동으로 구입한 로또와 자동으로 구입한 로또 모두를 가지고 있는 PurchasedLottos클래스 구현

indiveloper committed:
feat Rank Enum 구현

로또 당첨 정보를 가지고 있는 Rank Enum 구현

indiveloper committed:
feat getter 추가

money getter 추가

indiveloper committed:
feat String을 Int로 바꿔주는 Convertor 추가

입력 받은 String값을 int로 바꿔주는 Convertor 추가

indiveloper committed:
refactor Scanner.nextInt()를 Scanner.nextLine()으로 변경

Convertor를 이용해 Scanner.nextLine()으로 받은 값을 int 타입으로 만들어주는 방식으로 코드 수정

indiveloper committed:
feat 로또 결과 출력

로또 결과 출력 기능 추가.

indiveloper committed:
feat Main 클래스 추가 구현

Lotto를 진행하는 Main클래스 추가 구현

indiveloper committed:
style 사용하지 않는 import 제거 및 줄맞춤

indiveloper committed:
feat Lotto 번호 정렬을 위한 Comparable 구현

indiveloper committed:
refactor String을 StringBuilder로 변경

문자열의 연산이 이루어지는 부분을 String에서 StringBuilder로 변경

indiveloper committed:
refactor Collections.unmodifiableList를 이용해 반환값 수정

Collections.unmofiableList를 이용해 List를 변경하지 못하게 수정

indiveloper committed:
refactor ManualLottoGenerator 클래스 삭제에 따른 코드 수정

생성자를 호출해 Lotto 객체를 생성하는 역할만 하는 ManualLottoGenerator 클래스를 삭제하고 생성자를 직접 사용하는 것으로 코드 수정

indiveloper committed:
refactor getLotto()가 unmodifiable한 List를 반환함에 따른 코드 수정

indiveloper committed:
refactor LottoGenerator를 삭제하고 ShuffledNumberGenerator를 추가

AutoLottoGenerator와 ManualLottoGenerator를 삭제하고 lotto는 생성자로 생성함. 대신 ShuffledNumberGenerator를 이용해 랜덤한 6자리 숫자인 로또 번호만을 생성함.

indiveloper committed:
refactor Validator에서 수행하는 유효성 검사를 검사가 행해지는 클래스에서 하도록 수정

여러 클래스의 유효성 검사 기능을 Validator 클래스에 모아두었는데 이를 각자 클래스에서 수행하는 것으로 수정

indiveloper committed:
refactor Number와 NumberSet이 동일하게 사용하던 변수와 기능을 Number에서 캐싱하여 사용하는 것으로 수정

indiveloper committed:
refactor LottoGenerator 생성

LottoGenerator 인터페이스를 생성하고 AutoLottoGenerator와 ManualLottoGenerator에서 각각 LottoGenerator를 구현하여 자동/수동으로 로또를 생성하는 로직을 구현함.

indiveloper committed:
refactor for문으로 구현한 로직을 stream으로 수정

for문으로 구현한 로직을 stream.filter().findAy().orElseThrow()로 수정함.
로직에 변경은 없음.

indiveloper committed:
refactor Main에서 가지고 있던 메서드들을 각각의 객체로 분산

indiveloper committed:
refactor WinningLotto에서 순위를 판정하는 match기능 구현

indiveloper committed:
refactor 네이밍 변경

availablePurchaseTicketCount()에서 availablePurchaseCount()로 메서드 명 변경

indiveloper committed:
feat Number 클래스에 toString() 추가

indiveloper committed:
refactor 변수명 변경

~~Lottos였던 변수명을 ~~Lotto로 변경

indiveloper committed:
refactor getShuffledNumbers() 메서드 수정

기존에 numberPoll.subList()로 리턴해 사용하던 것을 새로운 List에 numberPoll.subList()를 넣어 리턴하는 것으로 수정.

indiveloper committed:
refactor input함수들에 조건 추가

checkManualLottoSize에 0장 이하일 경우 예외를 발생시키도록 조건 추가
printInputManualLottoMessage()에 수동 구입 장수가 0장 이하일 경우 메시지가 출력되지 않도록 조건 축가

indiveloper committed:
refactor 변수명 변경

Lottos에서 Lotto로 변수명 변경

indiveloper committed:
refactor 변수명 변경

Lottos에서 Lotto로 변수명 변경

indiveloper committed:
refactor LottoGenerator 클래스들 삭제

indiveloper committed:
refactor Main클래스 로직 수정

LottoGenerator를 없애고 Main 안에서 Lotto 생성을 수행

indiveloper committed:
refactor domain로직에 있던 출력문을 InputView 클래스의 메서드로 추출

indiveloper committed:
refactor Lotto 생성 부분 수정

makeAutoLotto()와 makeManualLotto()에서 수행하던 Lotto생성 기능을 Main메서드에서 담당하는 것으로 수정

indiveloper committed:
feat DB 생성

DB 생성

indiveloper committed:
refactor sql수정

round table의 pk를 id로 수정

indiveloper committed:
feat mysql dependency 추가

indiveloper committed:
feat MoneyDAO 클래스 생성

입력받은 로또 구입 금액을 입력 받아 DB에 INSERT 하는 기능 구현

indiveloper committed:
refactor DB 연결을 담당하는 ConnectionGenerator 생성

getConnection()을 static으로 선언해 util과 같이 사용할 수 있도록 ConnectionGenerator 클래스를 생성함

indiveloper committed:
feat LottoDAO 생성

Lotto를 db에 insert하는 lottoDAO 생성

indiveloper committed:
refactor MoneyDAO의 connection부분 제거

db의 connectino을 얻는 기능을 ConnectionGenerator클래스로 분리함에 따라 DAO클래스에서 Connection을 생성하는 부분을 삭제함.

indiveloper committed:
feat WinningLottoDAO 생성

당첨 번호를 DB에 입력하는 WinningLottoDAO 생성

indiveloper committed:
feat PrizeResultDAO 생성

당첨 번호를 DB에 입력하는 PrizeResultDAO 생성

indiveloper committed:
d

indiveloper committed:
flag before remove ajax

from now on, i will remove ajax because of changing of my code structure.

indiveloper committed:
flag changing code

commit this code even if still not completed because of my computer's instability

indiveloper committed:
feat lotto 웹, db 프로그램 구현

indiveloper committed:
test LottoDAO test 코드 추가

lottoDAO의 addLotto와 getPurchasedLotto의 테스트 코드 추가

indiveloper committed:
test WinningLottoDAO test 코드 추가

WinningLottoDAO의 addWinningLotto와 getWinningLotto의 테스트 코드 추가

indiveloper committed:
test ResultDAO test 코드 추가

ResultDAO의 test 코드 추가

indiveloper committed:
feat lotto 프로그램 html 파일 추가

soojinroh committed:
Merge branch 'soojinroh' into soojinroh

phs1116 commented on src/main/java/lotto/WebUILottoApplication.java:
```diff
@@ -9,10 +10,38 @@
+        });
+
+        get("/startGame", (req, res) -> {
+            Map<String, Object> model = new HashMap<>();
```
WebUILottoApplication에서 각 요청에 대한 구현을 상세히 알고 있는 것 같은데요,
Controller 객체 등으로 추상화해보면 어떨까요?

soojinroh replied:
WebUILottoApplication을 Controller라고 생각해서 구현 내용을 가지게 만들었던 것 같습니다. Controller 객체를 만들어서 WebUILottoApplication의 구현부를 추상화시켜보았습니다!

phs1116 commented on src/main/java/lotto/domain/dao/LottoDAO.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class LottoDAO {
```
현재 구현에 대해서 DAO는 Jdbc로 구현된 DAO이지요, 
각 서비스에서 DAO의 구현체를 직접 참조하지 않고, 인터페이스를 통해서 참조하도록 해보면 어떨까요?
그리고 인터페이스를 작성했을 경우에 얻는 잇점은 무엇일까요?

phs1116 replied:
DAO는 상태를 가지고 있지 않은 객체인데요, 매 요청마다 생성하지 않아도 될 것 같습니다.
싱글톤 패턴에 대해서 공부해서 적용해보면 어떨까요?

soojinroh replied:
싱글톤 패턴으로 DAOImpl을 만들고 인터페이스를 이용해 DAOImpl을 참조하는 것으로 수정했습니다!

DAO의 인터페이스를 만들면 persistence layer의 변경사항을 service layer가 알 필요 없이 사용할 수 있어 결합도를 낮출 수 있는 장점이 있는 것 같습니다. 어떤 DB를 사용하든 상관 없이 JDBC의 Interface를 이용해 DB를 사용할 수 있는 것과 비슷한 장점을 가지고 있는 것 같습니다.

싱글톤 패턴도 공부해서 적용해보았습니다.

피드백 감사합니다.

phs1116 commented on src/main/java/lotto/domain/dao/LottoDAO.java:
```diff
@@ -0,0 +1,68 @@
+
+    public void addLotto(final LottoDTO lottoDTO) {
+        try {
+            Connection con = ConnectionGenerator.getConnection();
```
`try-with-resources`에 대해서 공부해보고 적용해보면 좋을 것 같네요. :)

soojinroh replied:
try-with-resources를 공부해서 코드에 적용시켜 보았습니다. 이번 로또 미션에서 몇 부분은 Connection을 열고 close를 해주지 않는 실수를 해서 그런지 `try-with-sources`의 장점을 크게 느낄 수 있었습니다.

외부자원을 사용한 뒤 close하지 않는 실수를 방지할 수 있고 작성해야 할 코드 라인이 줄어들어 좋은 것 같습니다!

phs1116 commented on src/main/java/lotto/domain/dao/LottoDAO.java:
```diff
@@ -0,0 +1,68 @@
+        String query = "SELECT * FROM lotto WHERE round = ?";
+        PreparedStatement pstmt = ConnectionGenerator.getConnection().prepareStatement(query);
+        pstmt.setInt(1, round);
+        ResultSet rs = pstmt.executeQuery();
```
resultSet에 대해서도 꼭 close 해주세요!

soojinroh replied:
Connection을 열고 close하지 않은 부분을 전부 찾아 수정하였습니다!

phs1116 commented on src/main/java/lotto/domain/dao/LottoDAO.java:
```diff
@@ -0,0 +1,68 @@
+            pstmt.executeUpdate();
+            ConnectionGenerator.closeConnection(con);
+        } catch (SQLException e) {
+            System.out.println(e.getMessage());
```
예외를 이렇게 출력만하고 무시하는 경우에 대해서는 클라이언트 입장에서 정상으로 확인 될 가능성이 큰데요.
이펙티브 자바 예외 챕터 참고해보시면 좋을 것 같습니다. :)

그리고 예외의 핸들링에 대해 아래 블로그를 참조하여 공부하고 적용해보면 좋을 것 같네요. :)
http://wonwoo.ml/index.php/post/878

phs1116 replied:
그리고 System.out.println등 System.out을 통한 로그 출력에 대해서는 동기화에 따른 성능 문제 등으로 안티 패턴으로 간주됩니다. :)

soojinroh replied:
피드백 주신 것에 따라 이펙티브 자바 예외 챕터와 블로그 내용을 읽어보았습니다. 예외 처리가 중요하며 다양한 방법이 있다는 것, System.out으로 로그를 출력하는 것이 좋지 않다는 것 등을 알게 되었지만 아직 좋은 예외처리 방법에 대해서는 더 공부해야 할 것 같습니다. 일단은 System.out으로 예외처리 하던 것은 수정하였습니다.

예외처리는 항상 부수적인 것으로만 생각했는데 따로 공부할 주제로 잡아서 더 공부해보겠습니다!

phs1116 commented on src/main/java/lotto/domain/dao/ResultDAO.java:
```diff
@@ -0,0 +1,63 @@
+public class ResultDAO {
+
+    public void setResult(final ResultDTO resultDTO) throws SQLException  {
+        Connection con = ConnectionGenerator.getConnection();
```
DAO 클래스 전체적으로 커넥션 생성, executeUpdate등 쿼리 실행 등 공통적으로 try-catch 등 공통적으로 작업되는 내용이 많은데 이를 별도 템플릿 객체로 분리해보면 어떨까요?

soojinroh replied:
DAO 클래스에서 중복되는 코드를 템플릿 객체로 분리해서 구현해보았습니다. 페어의 도움을 많이 받아 구현한 것이라 충분히 스스로 고민해보지 못한 점도 있지만 템플릿으로 분리함으로써 생기는 장점을 충분히 느낄 수 있었습니다.

phs1116 commented on src/main/java/lotto/domain/dao/WinningLottoDAO.java:
```diff
@@ -0,0 +1,78 @@
+            return 1;
+        }
+
+        return rs.getInt("round") + NEXT_ROUND;
```
DAO 내에 비즈니스가 녹아있는 것 같네요.
DAO는 쿼리와 매핑에 대해서만 관리하고
새로운  Round를 가져오는 등 비즈니스 로직은 DAO가 아니라 서비스 레이어에서 처리하는게 적합합니다.

soojinroh replied:
DAO에서 getLatestRound()를 호출해서 가장 최근의 round를 가져오고, service layer에서 `+1`을 해주어 DAO에서 비즈니스 로직을 제거하였습니다. MVC패턴을 각각의 역할에 맞게 사용할 수 있도록 주의하겠습니다.

phs1116 commented on src/main/java/lotto/domain/dto/WinningLottoDTO.java:
```diff
@@ -0,0 +1,78 @@
+
+public class WinningLottoDTO {
+    private int round;
+    private Number firstNum;
```
각 DTO들도 기존 로또 도메인들과 같이 내부적으로 추상화하여 객체를 가지고 있을수 있지 않을까요? :)

phs1116 replied:
덧붙여서 각 DAO에 대해서는 DTO를 리턴하는게 아닌, 이전 단계에서 만든 Domain 객체로 매핑하도록 할 수 있을 것 같네요. :)

soojinroh replied:
생각 없이 습관적으로 DAO의 모든 리턴 값을 DTO로 만들었었는데 경우에 따라 DTO, VO를 사용하는 것이 좋겠다는 생각을 하게 되었습니다. 

DTO들의 필드를 로또 도메인과 같이 내부적으로 추상화시켰고 DAO의 리턴값도 수정해보았습니다! 피드백 감사합니다.

phs1116 commented on src/main/java/lotto/domain/utils/ConnectionGenerator.java:
```diff
@@ -0,0 +1,42 @@
+            System.out.println("정상적으로 연결되었습니다.");
+        } catch (SQLException e) {
+            System.err.println("연결 오류 : " + e.getMessage());
+            e.printStackTrace();
```
e.printStackTrace()에 대해서도 에러로 간주됩니다. :)
또한 해당 구현에 대해서는 NPE가 발생해서 상세 stackTrace나 로그내용을 찾지 않은 이상 원인을 바로 찾기 쉽지 않을 것 같은데요.
커넥션 생성에 실패시에는 별도의 예외를 던지는건 어떨까요?

soojinroh replied:
피드백 주신 내용 반영해서 Connection 생성 실패시에 별도의 예외를 만들어 던지는 것으로 수정했습니다.

예외처리에서 잘못 작성하고 있는 부분이 많은 것 같아 제대로 공부를 해야 할 것 같다는 필요성을 느끼고 있습니다. 피드백 감사합니다.

phs1116 reviewed (changes requested):
안녕하세요! 리뷰어 화투입니다.
기존 도메인 변경 없이 웹 적용 👍 
다만 몇가지 더 추상화 할 수 있는 점, 
그리고 예외 처리 부분에 대해서 피드백 추가했는데요.
참고 부탁드리겠습니다. :)
궁금하신 점 있으시면 언제든지 연락주세요!

indiveloper committed:
refactor WebUILottoApplication의 상세 구현부를 Controller객체로 추상화

WebUILottoApplication이 가지고 있던 상세 구현부를 Controller 클래스를 만들어 추상화키심.

indiveloper committed:
refactor DAO에서 Connection close 추가

Connection을 연결하고 close하지 않은 부분이 있어 Connection close를 추가

indiveloper committed:
refactor DAO에서 try-catch-fianlly를 try-with-resources로 변경

try-catch-finally로 jdbc를 연결해 사용하던 것을 try-with-resources로 변경해 사용.

indiveloper committed:
style 사용하지 않는 import, throws 제거 및 컨벤션 수정

indiveloper committed:
refactor JDBC Connect시 별도의 Exception을 던지는 것으로 수정

JDBCConnectException과 JDBCDriverLoadException을 만들어 JDBC 연결 시 해당 Exception을 던지는 것으로 수정하였음.

indiveloper committed:
style comma 추가

indiveloper committed:
refactor 피드백 적용

DAO Interface 생성
JDBC Template 생성
Controller 생성
SingleTon Pattern 적용해서 DAO 구현

indiveloper committed:
Merge remote-tracking branch 'origin/soojinroh' into soojinroh

phs1116 commented on src/main/java/lotto/domain/dao/JDBCTemplate.java:
```diff
@@ -0,0 +1,74 @@
+import java.util.List;
+import java.util.Map;
+
+public class JDBCTemplate {
```
JDBC Template 구현 👍

phs1116 commented on src/main/java/lotto/domain/dao/JDBCTemplate.java:
```diff
@@ -0,0 +1,74 @@
+        List<Map<String, String>> result = new ArrayList<>();
+
+        try (Connection con = Connector.getConnection()) {
+            PreparedStatement pstmt = getPreparedStatement(query, args, con);
```
PreparedStatement, ResultSet도 try-with-resources에 감싸서 닫아주는게 좋을 것 같네요. :)

phs1116 reviewed (approved):
안녕하세요! 리뷰어 화투입니다.
피드백 구현 감사합니다 전체적인 구조개선 👍 
이번 단계 머지하겠습니다.
피드백 한가지 추가했는데요, 한번 참고해보셔도 좋을 것 같습니다. :)
